lab_3/main.py

```diff
@@ -17,38 +17,87 @@ def tokenize_by_sentence(text: str) -> tuple:
          (('_', 'h', 'e', '_'), ('_', 'i', 's', '_'), ('_', 'h', 'a', 'p', 'p', 'y', '_'))
          )
     """
-    pass
+    checks = [
+        isinstance(text, str),
+        text != "",
+    ]
+
+    if not all(checks) or not any(letter.isalpha() for letter in text):
+        return ()
+
+    # generate indexes to split
+    split_indexes = []
+    for index in range(len(text[:-2])):
+        conditions = [not text[index].isalpha(),
+                      not text[index].isdigit(),
+                      text[index + 1] == " ",
+                      text[index + 2].isupper()]
+        if all(conditions):
+            split_indexes.append(index)
+
+    # split text
+    prev_split_point = 0
+    split_text = []
+    for index_split in split_indexes + [-1]:
+        sentence = "".join([word for word in text[prev_split_point:index_split].lower()
+                            if word.isalpha() or word == " "])
+        split_text.append(tuple([tuple(f"_{word}_") for word
+                                 in sentence.split()]))
+        prev_split_point = index_split + 2
+    return tuple(split_text)
 
 
 # 4
 class LetterStorage:
 
-    def __init__(self):
-        pass
+    def __init__(self, sentence=""):
+        self.storage = {}
 
     def _put_letter(self, letter: str) -> int:
         """
         Puts a letter into storage, assigns a unique id
         :param letter: a letter
         :return: 0 if succeeds, 1 if not
         """
-        pass
+
+        index_letter = 0
```

Comment:
If you define it here, the next time you try to add a letter, this will be the same `0` value.
You should have this index\id on the class level.

Reply:
yeah, you're right, thanks!